File: mobile-sdk/android/WebRTCManager.kt
package com.ooblex.sdk

import android.content.Context
import android.util.Log
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import org.json.JSONObject
import org.webrtc.*
import java.net.URL
import java.util.concurrent.Executors

/**
 * Manages WebRTC connections and video streaming
 */
class WebRTCManager(private val context: Context) {
    
    companion object {
        private const val TAG = "WebRTCManager"
    }
    
    // WebRTC components
    private var peerConnectionFactory: PeerConnectionFactory? = null
    private var peerConnection: PeerConnection? = null
    private var videoCapturer: CameraVideoCapturer? = null
    private var localVideoSource: VideoSource? = null
    private var localVideoTrack: VideoTrack? = null
    private var localAudioSource: AudioSource? = null
    private var localAudioTrack: AudioTrack? = null
    
    // Configuration
    private var serverURL: String = ""
    private var currentFrame: VideoFrame? = null
    private var frameCallback: ((VideoFrame) -> Unit)? = null
    
    // Executor for camera operations
    private val cameraExecutor = Executors.newSingleThreadExecutor()
    
    init {
        initializePeerConnectionFactory()
    }
    
    private fun initializePeerConnectionFactory() {
        val options = PeerConnectionFactory.InitializationOptions.builder(context)
            .setEnableInternalTracer(true)
            .createInitializationOptions()
        PeerConnectionFactory.initialize(options)
        
        val encoderFactory = DefaultVideoEncoderFactory(
            EglBase.create().eglBaseContext,
            true,
            true
        )
        val decoderFactory = DefaultVideoDecoderFactory(
            EglBase.create().eglBaseContext
        )
        
        peerConnectionFactory = PeerConnectionFactory.builder()
            .setVideoEncoderFactory(encoderFactory)
            .setVideoDecoderFactory(decoderFactory)
            .createPeerConnectionFactory()
    }
    
    fun configure(serverURL: String) {
        this.serverURL = serverURL
    }
    
    fun startCapture(
        cameraFacing: CameraFacing,
        resolution: VideoResolution,
        fps: Int,
        onFrameCaptured: (VideoFrame) -> Unit
    ) {
        frameCallback = onFrameCaptured
        
        val factory = peerConnectionFactory ?: return
        
        // Create video source
        localVideoSource = factory.createVideoSource(false)
        localVideoTrack = factory.createVideoTrack("video0", localVideoSource)
        localVideoTrack?.setEnabled(true)
        
        // Create audio source
        val audioConstraints = MediaConstraints()
        localAudioSource = factory.createAudioSource(audioConstraints)
        localAudioTrack = factory.createAudioTrack("audio0", localAudioSource)
        localAudioTrack?.setEnabled(true)
        
        // Set up camera capturer
        videoCapturer = createCameraCapturer(cameraFacing)
        videoCapturer?.initialize(
            SurfaceTextureHelper.create("CameraThread", EglBase.create().eglBaseContext),
            context,
            localVideoSource?.capturerObserver
        )
        
        videoCapturer?.startCapture(resolution.width, resolution.height, fps)
        
        // Set up frame observer
        localVideoSource?.capturerObserver?.let { observer ->
            localVideoSource?.setVideoProcessor(object : VideoProcessor {
                override fun onCapturerStarted(success: Boolean) {
                    Log.d(TAG, "Capturer started: $success")
                }
                
                override fun onCapturerStopped() {
                    Log.d(TAG, "Capturer stopped")
                }
                
                override fun onFrameCaptured(frame: VideoFrame) {
                    currentFrame = frame
                    frameCallback?.invoke(frame)
                }
                
                override fun setSink(sink: VideoSink?) {
                    // Not used in this implementation
                }
            })
        }
    }
    
    fun stopCapture() {
        videoCapturer?.stopCapture()
        videoCapturer?.dispose()
        videoCapturer = null
        
        localVideoTrack?.setEnabled(false)
        localVideoTrack?.dispose()
        localVideoTrack = null
        
        localVideoSource?.dispose()
        localVideoSource = null
        
        localAudioTrack?.setEnabled(false)
        localAudioTrack?.dispose()
        localAudioTrack = null
        
        localAudioSource?.dispose()
        localAudioSource = null
    }
    
    suspend fun connect() = withContext(Dispatchers.IO) {
        val factory = peerConnectionFactory ?: throw Exception("PeerConnectionFactory not initialized")
        
        val iceServers = listOf(
            PeerConnection.IceServer.builder("stun:stun.l.google.com:19302").createIceServer()
        )
        
        val rtcConfig = PeerConnection.RTCConfiguration(iceServers).apply {
            bundlePolicy = PeerConnection.BundlePolicy.MAXBUNDLE
            rtcpMuxPolicy = PeerConnection.RtcpMuxPolicy.REQUIRE
            tcpCandidatePolicy = PeerConnection.TcpCandidatePolicy.DISABLED
            candidateNetworkPolicy = PeerConnection.CandidateNetworkPolicy.ALL
            continualGatheringPolicy = PeerConnection.ContinualGatheringPolicy.GATHER_CONTINUALLY
        }
        
        peerConnection = factory.createPeerConnection(
            rtcConfig,
            object : PeerConnection.Observer {
                override fun onSignalingChange(state: PeerConnection.SignalingState) {
                    Log.d(TAG, "Signaling state changed: $state")
                }
                
                override fun onIceConnectionChange(state: PeerConnection.IceConnectionState) {
                    Log.d(TAG, "ICE connection state changed: $state")
                }
                
                override fun onIceConnectionReceivingChange(receiving: Boolean) {
                    Log.d(TAG, "ICE connection receiving changed: $receiving")
                }
                
                override fun onIceGatheringChange(state: PeerConnection.IceGatheringState) {
                    Log.d(TAG, "ICE gathering state changed: $state")
                }
                
                override fun onIceCandidate(candidate: IceCandidate) {
                    sendIceCandidateToServer(candidate)
                }
                
                override fun onIceCandidatesRemoved(candidates: Array<out IceCandidate>) {
                    Log.d(TAG, "ICE candidates removed")
                }
                
                override fun onAddStream(stream: MediaStream) {
                    Log.d(TAG, "Stream added")
                }
                
                override fun onRemoveStream(stream: MediaStream) {
                    Log.d(TAG, "Stream removed")
                }
                
                override fun onDataChannel(dataChannel: DataChannel) {
                    Log.d(TAG, "Data channel created")
                }
                
                override fun onRenegotiationNeeded() {
                    Log.d(TAG, "Renegotiation needed")
                }
                
                override fun onAddTrack(receiver: RtpReceiver, streams: Array<out MediaStream>) {
                    Log.d(TAG, "Track added")
                }
            }
        )
        
        // Add tracks to peer connection
        localVideoTrack?.let { track ->
            peerConnection?.addTrack(track, listOf("stream0"))
        }
        localAudioTrack?.let { track ->
            peerConnection?.addTrack(track, listOf("stream0"))
        }
        
        // Create and send offer
        createAndSendOffer()
    }
    
    fun disconnect() {
        peerConnection?.close()
        peerConnection = null
    }
    
    fun getLocalVideoTrack(): VideoTrack? = localVideoTrack
    
    fun getProcessedVideoTrack(): VideoTrack? {
        // In a real implementation, this would return the processed video track from the server
        return localVideoTrack
    }
    
    fun getCurrentFrame(): VideoFrame? = currentFrame
    
    fun attachRenderer(renderer: SurfaceViewRenderer) {
        localVideoTrack?.addSink(renderer)
    }
    
    fun detachRenderer(renderer: SurfaceViewRenderer) {
        localVideoTrack?.removeSink(renderer)
    }
    
    fun release() {
        stopCapture()
        disconnect()
        peerConnectionFactory?.dispose()
        peerConnectionFactory = null
        cameraExecutor.shutdown()
    }
    
    private fun createCameraCapturer(cameraFacing: CameraFacing): CameraVideoCapturer? {
        val enumerator = Camera2Enumerator(context)
        
        val deviceNames = enumerator.deviceNames
        
        // Try to find the requested camera
        for (deviceName in deviceNames) {
            if (cameraFacing == CameraFacing.FRONT && enumerator.isFrontFacing(deviceName)) {
                return enumerator.createCapturer(deviceName, null)
            } else if (cameraFacing == CameraFacing.BACK && enumerator.isBackFacing(deviceName)) {
                return enumerator.createCapturer(deviceName, null)
            }
        }
        
        // Fall back to first available camera
        return if (deviceNames.isNotEmpty()) {
            enumerator.createCapturer(deviceNames[0], null)
        } else {
            null
        }
    }
    
    private suspend fun createAndSendOffer() = withContext(Dispatchers.IO) {
        val constraints = MediaConstraints().apply {
            mandatory.add(MediaConstraints.KeyValuePair("OfferToReceiveVideo", "true"))
            mandatory.add(MediaConstraints.KeyValuePair("OfferToReceiveAudio", "false"))
        }
        
        peerConnection?.createOffer(object : SdpObserver {
            override fun onCreateSuccess(sessionDescription: SessionDescription) {
                peerConnection?.setLocalDescription(object : SdpObserver {
                    override fun onCreateSuccess(p0: SessionDescription?) {}
                    override fun onSetSuccess() {
                        sendOfferToServer(sessionDescription)
                    }
                    override fun onCreateFailure(p0: String?) {}
                    override fun onSetFailure(p0: String?) {}
                }, sessionDescription)
            }
            
            override fun onSetSuccess() {}
            override fun onCreateFailure(error: String?) {
                Log.e(TAG, "Failed to create offer: $error")
            }
            override fun onSetFailure(error: String?) {}
        }, constraints)
    }
    
    private fun sendOfferToServer(sdp: SessionDescription) {
        cameraExecutor.execute {
            try {
                val url = URL("$serverURL/api/webrtc/offer")
                val connection = url.openConnection() as java.net.HttpURLConnection
                connection.requestMethod = "POST"
                connection.setRequestProperty("Content-Type", "application/json")
                connection.doOutput = true
                
                val json = JSONObject().apply {
                    put("sdp", sdp.description)
                    put("type", sdp.type.canonicalForm())
                }
                
                connection.outputStream.use { os ->
                    os.write(json.toString().toByteArray())
                }
                
                if (connection.responseCode == 200) {
                    val response = connection.inputStream.bufferedReader().use { it.readText() }
                    val responseJson = JSONObject(response)
                    
                    val answerSdp = SessionDescription(
                        SessionDescription.Type.fromCanonicalForm(responseJson.getString("type")),
                        responseJson.getString("sdp")
                    )
                    
                    peerConnection?.setRemoteDescription(object : SdpObserver {
                        override fun onCreateSuccess(p0: SessionDescription?) {}
                        override fun onSetSuccess() {
                            Log.d(TAG, "Remote description set successfully")
                        }
                        override fun onCreateFailure(p0: String?) {}
                        override fun onSetFailure(error: String?) {
                            Log.e(TAG, "Failed to set remote description: $error")
                        }
                    }, answerSdp)
                }
            } catch (e: Exception) {
                Log.e(TAG, "Failed to send offer to server", e)
            }
        }
    }
    
    private fun sendIceCandidateToServer(candidate: IceCandidate) {
        cameraExecutor.execute {
            try {
                val url = URL("$serverURL/api/webrtc/ice")
                val connection = url.openConnection() as java.net.HttpURLConnection
                connection.requestMethod = "POST"
                connection.setRequestProperty("Content-Type", "application/json")
                connection.doOutput = true
                
                val json = JSONObject().apply {
                    put("candidate", candidate.sdp)
                    put("sdpMLineIndex", candidate.sdpMLineIndex)
                    put("sdpMid", candidate.sdpMid)
                }
                
                connection.outputStream.use { os ->
                    os.write(json.toString().toByteArray())
                }
                
                if (connection.responseCode != 200) {
                    Log.e(TAG, "Failed to send ICE candidate: ${connection.responseCode}")
                }
            } catch (e: Exception) {
                Log.e(TAG, "Failed to send ICE candidate to server", e)
            }
        }
    }
}

File: mobile-sdk/ios/WebRTCManager.swift
import Foundation
import WebRTC

/// Manages WebRTC connections and video streaming
class WebRTCManager: NSObject {
    
    // MARK: - Properties
    
    private var peerConnectionFactory: RTCPeerConnectionFactory!
    private var peerConnection: RTCPeerConnection?
    private var localVideoSource: RTCVideoSource?
    private var videoCapturer: RTCCameraVideoCapturer?
    private var localVideoTrack: RTCVideoTrack?
    private var serverURL: String = ""
    
    var currentFrame: RTCVideoFrame?
    var onFrameCaptured: ((RTCVideoFrame) -> Void)?
    
    // MARK: - Initialization
    
    override init() {
        super.init()
        initializeWebRTC()
    }
    
    private func initializeWebRTC() {
        RTCInitializeSSL()
        
        let videoEncoderFactory = RTCDefaultVideoEncoderFactory()
        let videoDecoderFactory = RTCDefaultVideoDecoderFactory()
        
        peerConnectionFactory = RTCPeerConnectionFactory(
            encoderFactory: videoEncoderFactory,
            decoderFactory: videoDecoderFactory
        )
    }
    
    // MARK: - Configuration
    
    func configure(serverURL: String) {
        self.serverURL = serverURL
    }
    
    // MARK: - Video Capture
    
    func startCapture(cameraPosition: AVCaptureDevice.Position,
                     resolution: VideoResolution,
                     fps: Int) throws {
        localVideoSource = peerConnectionFactory.videoSource()
        videoCapturer = RTCCameraVideoCapturer(delegate: localVideoSource!)
        
        let devices = RTCCameraVideoCapturer.captureDevices()
        guard let device = devices.first(where: { $0.position == cameraPosition }) else {
            throw WebRTCError.noCameraAvailable
        }
        
        let format = selectFormat(for: device, targetWidth: resolution.width, targetHeight: resolution.height)
        let fps = selectFPS(for: format, target: fps)
        
        videoCapturer?.startCapture(with: device, format: format, fps: fps)
        
        localVideoTrack = peerConnectionFactory.videoTrack(with: localVideoSource!, trackId: "video0")
        localVideoTrack?.isEnabled = true
        
        // Set up frame capture
        localVideoSource?.adaptOutputFormat(
            toWidth: resolution.width,
            height: resolution.height,
            fps: Int32(fps)
        )
    }
    
    func stopCapture() {
        videoCapturer?.stopCapture()
        localVideoTrack?.isEnabled = false
        localVideoTrack = nil
        localVideoSource = nil
        videoCapturer = nil
    }
    
    // MARK: - Connection Management
    
    func connect(completion: @escaping (Result<Void, Error>) -> Void) {
        createPeerConnection { [weak self] result in
            switch result {
            case .success:
                self?.createOffer(completion: completion)
            case .failure(let error):
                completion(.failure(error))
            }
        }
    }
    
    func disconnect() {
        peerConnection?.close()
        peerConnection = nil
    }
    
    // MARK: - Private Methods
    
    private func createPeerConnection(completion: @escaping (Result<Void, Error>) -> Void) {
        let config = RTCConfiguration()
        config.iceServers = [RTCIceServer(urlStrings: ["stun:stun.l.google.com:19302"])]
        config.sdpSemantics = .unifiedPlan
        config.continualGatheringPolicy = .gatherContinually
        
        let constraints = RTCMediaConstraints(
            mandatoryConstraints: nil,
            optionalConstraints: ["DtlsSrtpKeyAgreement": kRTCMediaConstraintsValueTrue]
        )
        
        guard let pc = peerConnectionFactory.peerConnection(
            with: config,
            constraints: constraints,
            delegate: self
        ) else {
            completion(.failure(WebRTCError.peerConnectionFailed))
            return
        }
        
        peerConnection = pc
        
        if let localVideoTrack = localVideoTrack {
            pc.add(localVideoTrack, streamIds: ["stream0"])
        }
        
        completion(.success(()))
    }
    
    private func createOffer(completion: @escaping (Result<Void, Error>) -> Void) {
        let constraints = RTCMediaConstraints(
            mandatoryConstraints: [
                kRTCMediaConstraintsOfferToReceiveAudio: kRTCMediaConstraintsValueFalse,
                kRTCMediaConstraintsOfferToReceiveVideo: kRTCMediaConstraintsValueTrue
            ],
            optionalConstraints: nil
        )
        
        peerConnection?.offer(for: constraints) { [weak self] sdp, error in
            guard let self = self, let sdp = sdp else {
                completion(.failure(error ?? WebRTCError.offerCreationFailed))
                return
            }
            
            self.peerConnection?.setLocalDescription(sdp) { error in
                if let error = error {
                    completion(.failure(error))
                } else {
                    self.sendOfferToServer(sdp: sdp, completion: completion)
                }
            }
        }
    }
    
    private func sendOfferToServer(sdp: RTCSessionDescription, completion: @escaping (Result<Void, Error>) -> Void) {
        // Send SDP offer to Ooblex server
        let url = URL(string: "\(serverURL)/api/webrtc/offer")!
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        
        let body: [String: Any] = [
            "sdp": sdp.sdp,
            "type": RTCSessionDescription.string(for: sdp.type)
        ]
        
        request.httpBody = try? JSONSerialization.data(withJSONObject: body)
        
        URLSession.shared.dataTask(with: request) { [weak self] data, response, error in
            if let error = error {
                completion(.failure(error))
                return
            }
            
            guard let data = data,
                  let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
                  let answerSDP = json["sdp"] as? String,
                  let answerTypeString = json["type"] as? String,
                  let answerType = RTCSessionDescription.type(for: answerTypeString) else {
                completion(.failure(WebRTCError.invalidServerResponse))
                return
            }
            
            let answer = RTCSessionDescription(type: answerType, sdp: answerSDP)
            self?.peerConnection?.setRemoteDescription(answer) { error in
                if let error = error {
                    completion(.failure(error))
                } else {
                    completion(.success(()))
                }
            }
        }.resume()
    }
    
    private func selectFormat(for device: AVCaptureDevice, targetWidth: Int32, targetHeight: Int32) -> AVCaptureDevice.Format {
        let formats = RTCCameraVideoCapturer.supportedFormats(for: device)
        var selectedFormat = formats[0]
        var currentDiff = Int32.max
        
        for format in formats {
            let dimension = CMVideoFormatDescriptionGetDimensions(format.formatDescription)
            let diff = abs(dimension.width - targetWidth) + abs(dimension.height - targetHeight)
            if diff < currentDiff {
                selectedFormat = format
                currentDiff = diff
            }
        }
        
        return selectedFormat
    }
    
    private func selectFPS(for format: AVCaptureDevice.Format, target: Int) -> Int {
        var maxFPS = 0
        for range in format.videoSupportedFrameRateRanges {
            maxFPS = Int(max(maxFPS, Int(range.maxFrameRate)))
        }
        return min(maxFPS, target)
    }
}

// MARK: - RTCPeerConnectionDelegate

extension WebRTCManager: RTCPeerConnectionDelegate {
    func peerConnection(_ peerConnection: RTCPeerConnection, didChange stateChanged: RTCSignalingState) {
        print("Signaling state changed: \(stateChanged)")
    }
    
    func peerConnection(_ peerConnection: RTCPeerConnection, didAdd stream: RTCMediaStream) {
        print("Stream added")
    }
    
    func peerConnection(_ peerConnection: RTCPeerConnection, didRemove stream: RTCMediaStream) {
        print("Stream removed")
    }
    
    func peerConnectionShouldNegotiate(_ peerConnection: RTCPeerConnection) {
        print("Negotiation needed")
    }
    
    func peerConnection(_ peerConnection: RTCPeerConnection, didChange newState: RTCIceConnectionState) {
        print("ICE connection state changed: \(newState)")
    }
    
    func peerConnection(_ peerConnection: RTCPeerConnection, didChange newState: RTCIceGatheringState) {
        print("ICE gathering state changed: \(newState)")
    }
    
    func peerConnection(_ peerConnection: RTCPeerConnection, didGenerate candidate: RTCIceCandidate) {
        // Send ICE candidate to server
        sendIceCandidateToServer(candidate: candidate)
    }
    
    func peerConnection(_ peerConnection: RTCPeerConnection, didRemove candidates: [RTCIceCandidate]) {
        print("ICE candidates removed")
    }
    
    func peerConnection(_ peerConnection: RTCPeerConnection, didOpen dataChannel: RTCDataChannel) {
        print("Data channel opened")
    }
    
    private func sendIceCandidateToServer(candidate: RTCIceCandidate) {
        let url = URL(string: "\(serverURL)/api/webrtc/ice")!
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        
        let body: [String: Any] = [
            "candidate": candidate.sdp,
            "sdpMLineIndex": candidate.sdpMLineIndex,
            "sdpMid": candidate.sdpMid ?? ""
        ]
        
        request.httpBody = try? JSONSerialization.data(withJSONObject: body)
        URLSession.shared.dataTask(with: request).resume()
    }
}

// MARK: - RTCVideoCapturerDelegate

extension WebRTCManager: RTCVideoCapturerDelegate {
    func capturer(_ capturer: RTCVideoCapturer, didCapture frame: RTCVideoFrame) {
        currentFrame = frame
        onFrameCaptured?(frame)
    }
}

// MARK: - Errors

enum WebRTCError: Error, LocalizedError {
    case noCameraAvailable
    case peerConnectionFailed
    case offerCreationFailed
    case invalidServerResponse
    
    var errorDescription: String? {
        switch self {
        case .noCameraAvailable:
            return "No camera available"
        case .peerConnectionFailed:
            return "Failed to create peer connection"
        case .offerCreationFailed:
            return "Failed to create SDP offer"
        case .invalidServerResponse:
            return "Invalid response from server"
        }
    }
}